libkbfs/block_retrieval_queue.go

```diff
@@ -163,8 +169,84 @@ func (brq *blockRetrievalQueue) notifyWorker() {
 	}
 }
 
+func (brq *blockRetrievalQueue) CacheAndPrefetch(ptr BlockPointer, block Block,
+	kmd KeyMetadata, priority int, lifetime BlockCacheLifetime,
+	hasPrefetched bool) (err error) {
+	dbc := brq.config.DiskBlockCache()
+	if dbc != nil {
+		go dbc.UpdateMetadata(context.TODO(), ptr.ID)
+	}
+	if hasPrefetched {
+		return brq.config.BlockCache().PutWithPrefetch(ptr, kmd.TlfID(), block,
+			lifetime, true)
+	}
+	if priority < defaultOnDemandRequestPriority {
+		// Only on-demand or higher priority requests can trigger prefetches.
+		return brq.config.BlockCache().PutWithPrefetch(ptr, kmd.TlfID(), block,
+			lifetime, false)
+	}
+	// We must let the cache know at this point that we've prefetched.
+	// 1) To prevent any other Gets from prefetching.
+	// 2) To prevent prefetching if a cache Put fails, since prefetching if
+	//	  only useful when combined with the cache.
+	err = brq.config.BlockCache().PutWithPrefetch(ptr, kmd.TlfID(), block,
+		lifetime, true)
+	if _, isCacheFullError := err.(cachePutCacheFullError); isCacheFullError {
+		brq.log.CDebugf(context.TODO(), "Skipping prefetch because the cache "+
+			"is full")
+		return err
+	}
```

Comment:
Probably good to add an `else if err != nil` clause that prints the other errors so they don't get completely lost.